# Convolutional Neural Network (CNN) Example - CIFAR-10

This notebook demonstrates training a simple CNN on the CIFAR-10 dataset using TensorFlow/Keras.

In [1]:
import tensorflow as tf
from tensorflow.keras import datasets, layers, models
import matplotlib.pyplot as plt

# Load and preprocess the CIFAR-10 dataset
(train_images, train_labels), (test_images, test_labels) = datasets.cifar10.load_data()

train_images, test_images = train_images / 255.0, test_images / 255.0

class_names = ['airplane', 'automobile', 'bird', 'cat', 'deer',
               'dog', 'frog', 'horse', 'ship', 'truck']

170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 3s 0us/step


In [2]:
# Build the CNN model
model = models.Sequential([
    layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.MaxPooling2D((2, 2)),
    layers.Conv2D(64, (3, 3), activation='relu'),
    layers.Flatten(),
    layers.Dense(64, activation='relu'),
    layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [3]:
# Train the model
history = model.fit(train_images, train_labels, epochs=5,
                    validation_data=(test_images, test_labels))

Epoch 1/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 74s 46ms/step - accuracy: 0.3527 - loss: 1.7478 - val_accuracy: 0.5560 - val_loss: 1.2608
Epoch 2/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 80s 45ms/step - accuracy: 0.5903 - loss: 1.1461 - val_accuracy: 0.6290 - val_loss: 1.0496
Epoch 3/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 71s 46ms/step - accuracy: 0.6514 - loss: 0.9878 - val_accuracy: 0.6625 - val_loss: 0.9637
Epoch 4/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 81s 45ms/step - accuracy: 0.6930 - loss: 0.8821 - val_accuracy: 0.6931 - val_loss: 0.8819
Epoch 5/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 82s 45ms/step - accuracy: 0.7149 - loss: 0.8075 - val_accuracy: 0.6898 - val_loss: 0.9014


In [4]:
# Evaluate the model
test_loss, test_acc = model.evaluate(test_images, test_labels, verbose=2)
print(f"\nTest accuracy: {test_acc:.4f}")

313/313 - 5s - 15ms/step - accuracy: 0.6898 - loss: 0.9014

Test accuracy: 0.6898


The CNN model trained on CIFAR-10 reached about 71% training accuracy and 67% validation accuracy after 5 epochs, showing it learned well but with slight overfitting. The validation loss stabilized near 0.96, suggesting the model generalizes fairly well but could improve with more tuning or regularization.
